pandas/core/groupby/generic.py

```diff
@@ -1428,6 +1428,12 @@ def transform(self, func, *args, **kwargs):
         if not result.columns.equals(obj.columns):
             return self._transform_general(func, *args, **kwargs)
 
+        # GH 30918
+        if len(result) != self.ngroups:
```

Comment:
This doesn't seem like the right fix; do you know where the error occurs in `transform_fast`?

Reply:
Yes. It is with this line:
```python
res = algorithms.take_1d(result.iloc[:, i].values, ids)
```
This line tries to broadcast aggregated results to the entire group, which is necessary for other use cases of `transform`. Example:
```python
>>> df = pd.DataFrame([[1, 1], [1, 2], [2, 3], [2, 4]], columns=['a', 'b'])
>>> df
   a  b
0  1  1
1  1  2
2  2  3
3  2  4
>>> df.groupby('a').transform('mean')
     b
0  1.5
1  1.5
2  3.5
3  3.5
```
Inside the transform method result is
```python
result = df.groupby('a').mean()  # result = getattr(self, func)(*args, **kwargs)
>>> result
     b
a
1  1.5
2  3.5
```
The `_transform_fast` method then broadcast this result with that`algorithms.take_1d` call.

I think this is not the problem with `_transform_fast`, so that's why I implemented the fix outside. There are a few other if statements that server similar purposes too.

Reply:
In the docstring of `_transform_fast` we have this:
```python
"""
Fast transform path for aggregations
"""
```
This means it is meant only for aggregation `func` such as `mean`, `first`, `sum`.
The name `_transform_fast` is a bit confusing as it is meant for a different purpose, not just that it is faster.

I added further conditions before calling `_transform_fast` so that it is only used as intended. This resolves the error and also cleaned up the code a little bit.